A flexible particle binning plugin for particles properties.

ikbuibui opened the pull request:
Users can 
- Define their own axes
 - Define their own quantity which is binned
 - Choose which species which are used for the binning
 - Choose how frequently they want the binning to be executed
 - Choose if the binning should be time averaging or normalized by bin volume
 - Write custom output to file, for example other quantites related to the simulation which the user is interested in
 - Execute multiple binnings at the same time

Intends to solve #4513

ikbuibui committed:
A flexible particle binning plugin
Binning for particle properties and users can
- Define their own axes
- Define their own quantity which is binned
- Choose which species which are used for the binning
- Choose how frequently they want the binning to be executed
- Choose if the binning should be time averaging or normalized by bin volume
- Write custom output to file, for example other quantites related to the simulation which the user is interested in
- Execute multiple binnings at the same time

ikbuibui committed:
Add STLTuple implementation

psychocoderHPC commented on include/picongpu/param/binningSetup.param:
```diff
@@ -0,0 +1,198 @@
+            }
+        }
+    } // namespace plugins::binning
+} // namespace picongpu
```
missing EOF

ikbuibui replied:
Added

psychocoderHPC commented on include/picongpu/plugins/binning/utility.hpp:
```diff
@@ -0,0 +1,79 @@
+                pmacc::memory::tuple::Tuple<TArgs...>&& t,
+                std::index_sequence<Is...>)
+            {
+                // TODO give Is as a param to f, so compiler has knowledge
```
please use `@todo`, docygen will create  a todo list with all todo's in the cde

ikbuibui replied:
Switched

psychocoderHPC commented on include/picongpu/plugins/binning/utility.hpp:
```diff
@@ -0,0 +1,79 @@
+        }
+
+    } // namespace plugins::binning
+} // namespace picongpu
```
missing EOF

ikbuibui committed:
Change and use pmacc tuple
Only do the test binning if electrons exist in the sim

psychocoderHPC commented on include/picongpu/plugins/binning/Axis.hpp:
```diff
@@ -0,0 +1,317 @@
+            class Range
+            {
+            public:
+                T_Data min; ///< Minimum of binning range
```
Please move comments always above the variable and use `/**` to start the doxygen comment.
Could you please add to the comment if the value is in PIC or SI units.

ikbuibui replied:
Changed comment style and added info that it is in SI units

psychocoderHPC commented on include/picongpu/plugins/binning/Axis.hpp:
```diff
@@ -0,0 +1,317 @@
+            class AxisSplitting
+            {
+            public:
+                Range<T_Data> m_range; ///< Range object
```
Same as above, please sue doxygen and add a comment about the unit.

psychocoderHPC commented on include/picongpu/plugins/binning/Binner.hpp:
```diff
@@ -0,0 +1,319 @@
+        template<uint32_t blockSize, uint32_t nAxes>
+        struct BinNormalizationKernel
+        {
+            using result_type = void;
```
should be named `ResultType` or `ValueType` to follow our other naming schemas.

ikbuibui replied:
switched to ResultType

psychocoderHPC commented on include/picongpu/plugins/binning/Binner.hpp:
```diff
@@ -0,0 +1,319 @@
+                 */
+
+                this->histBuffer = std::make_unique<HostDeviceBuffer<TDepositedQuantity, 1>>(
+                    binningData.axisExtentsND.productOfComponents()); // TODO for auto n_bins. allocate full size
```
please move the comment into the comment block above.

ikbuibui replied:
Moved

psychocoderHPC commented on include/picongpu/plugins/binning/Binner.hpp:
```diff
@@ -0,0 +1,319 @@
+        };
+
+    } // namespace plugins::binning
+} // namespace picongpu
```
missing EOF

psychocoderHPC commented on include/picongpu/plugins/binning/BinningCreator.hpp:
```diff
@@ -0,0 +1,89 @@
+            }
+        };
+    } // namespace plugins::binning
+} // namespace picongpu
```
missing EOF

psychocoderHPC commented on include/picongpu/plugins/binning/BinningData.hpp:
```diff
@@ -0,0 +1,95 @@
+            }
+        };
+    }; // namespace plugins::binning
+} // namespace picongpu
```
missing EOF

psychocoderHPC commented on include/picongpu/plugins/binning/Axis.hpp:
```diff
@@ -0,0 +1,317 @@
+                    T min, max; /**< Min and max values in the range of the binning. Values outside this range are
+                                placed in overflow bins */
+                    uint32_t nBins; ///< Number of bins in range
+                    using ScalingType = std::
```
please add a comment that we use a type that avoids precision loss.

ikbuibui replied:
added

franzpoeschel commented on include/picongpu/plugins/binning/DomainInfo.hpp:
```diff
@@ -0,0 +1,54 @@
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
```
I'm having some trouble compiling this plugin at the moment since some `.hpp` files (this one, but also others like `UnitConversion.hpp`) have no `#include` directives, so only a very specific include order will work.
Can you give a short guide on how to compile and run a working instance of this plugin?

psychocoderHPC replied:
could you please post the compiler error log that we can trace from where the error starts.

franzpoeschel replied:
It's not only one file that is affected by this:

```
[ 97%] Building CUDA object CMakeFiles/picongpu.dir/versionFormat.cpp.o                                                                                                                                                                                                                                                    
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: the global scope has no "openPMD"                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: too few arguments for class template "std::map"                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: expected an identifier                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(74): error: the global scope has no "openPMD"                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(74): error: expected a ";"                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name  
```

When I fix the includes here, the next include error is in `DomainInfo.hpp`.

Error triggered by adding `extensionParam.loader` with below content to LaserWakefield example and trying to compile it:

```cpp
#pragma once
#include "picongpu/param/binningSetup.param"
```

ikbuibui replied:
The param file actually doesn't need to be added in extensionParam.loader, as the include is internally dealt with in the plugin.

psychocoderHPC replied:
This plugin is different compared to our old plugins. We kept the configuration in the param folder but this file is not allowed to be included from `param.loader` or the extension.
The plugin will include the param files. This will also solve typical cyclic includes we have in PIConGPU.

ikbuibui replied:
Also, could you pull again before trying again :smile:

franzpoeschel replied:
I pulled again. How do I run this then? There is nothing in `picongpu --help` except for the single line `Binning Dispatcher: A flexible binning plugin!:`

franzpoeschel replied:
Or are this the `--e_energyHistogram.*` parameters?

ikbuibui replied:
So there is no runtime configuration, and to run the plugin, the user config goes into `picongpu/param/binningSetup.param` . There is an existing example in that file which you can simply call in the `getBinning()` function as `binningExample(binningCreator)`

franzpoeschel replied:
Ah. I also had a simulation where the `if constexpr` part in the example apparently evaluated to `false`, so nothing happened. I have it running now.

ikbuibui replied:
I realized that it was always evaluating to false because i was using `mpl::true_` instead of `mp11::mp_true` :sweat_smile:

franzpoeschel commented on include/picongpu/param/binningSetup.param:
```diff
@@ -0,0 +1,216 @@
+                {
+                    mesh.setAttribute("p_nbins", ax_py.getNBins());
+                    // mesh.setAttribute("p_min", ax_py.getMin()); // This line throws an error in particle
+                    // calorimeter mesh.setAttribute("p_max", ax_py.getMax()); // This line throws an error in
```
Can you try splitting this on multiple lines? I'd like to see if the error comes from `getMin()` or from `setAttribute()`. Given that it's an error somewhere inside CUDA, it's most likely the `getMin()` and `getMax()` thing. I can imagine that copying `ax_py` into the closure might at this point not be enough to keep the underlying data alive/valid.

ikbuibui replied:
The copy into the closure works, and writes output as expected. 
I tried recreating the error, and though I was struggling with it for a long time, I can no longer trigger it. 
I first thought maybe the error went away when I renamed the `binningSetup` from `.hpp` to `.param`, but that's not the case. So, I will for now ignore this until it breaks in the future. I also cleaned up some includes and namespaces, so that might also have helped at some point.  
Sorry about the hunt that went nowhere, but thanks for being the first person to pull and compile the plugin :smile:, I've updated the documentation with the things learnt from your first use.

franzpoeschel commented on include/picongpu/plugins/binning/WriteHist.hpp:
```diff
@@ -0,0 +1,171 @@
+                record.setUnitSI(get_conversion_factor(outputUnits));
+
+                record.resetDataset({::openPMD::determineDatatype<Type>(), histExtent});
+                std::shared_ptr<Type> data(hReducedBuffer.getBasePointer(), [](auto const&) {});
+                record.storeChunk<Type>(data, histOffset, histExtent);
```
```suggestion
#if OPENPMDAPI_VERSION_GE(0, 15, 0)
                ::openPMD::UniquePtrWithLambda<Type> data(
                    hReducedBuffer.getBasePointer(),
                    [bfr = std::move(hReducedBuffer)](auto const*) {});
#else
                std::shared_ptr<Type> data(hReducedBuffer.getBasePointer(), [](auto const&) {});
#endif
                record.storeChunk<Type>(std::move(data), histOffset, histExtent);
```

This can help save memory in the BP5 engine when using a somewhat recent installation of openPMD-api. 
I'll need to check the exact version again and the `hReducedBuffer` would then need to be passed into this whole function by move instead of by reference.

If I don't forget it, I'll prepare a commit for this next week.

franzpoeschel replied:
Hm, I get errors from this. Need to look more closely next week.

psychocoderHPC replied:
Thanks for looking into it. We will keep this PR the code as it is and your optimization should go into the code a separate optimization.
I assume your optimization should be applied to other parts of PIConGPU too.

psychocoderHPC replied:
IMO `std::move(hReducedBuffer)` is not allowed because `hReducedBuffer` is managed by PIConGPU and the plugin is reusing the buffer. If you move the object into the lambda it would mean you are not allowed to touch the buffer anymore.

franzpoeschel replied:
> I assume your optimization should be applied to other parts of PIConGPU too.

No, most other parts of PIConGPU already use `storeChunkSpan()` from `include/picongpu/plugins/common/openPMDVersion.def` which is a better optimization than this one, but would require heavy restructuring here.

franzpoeschel replied:
> IMO `std::move(hReducedBuffer)` is not allowed because `hReducedBuffer` is managed by PIConGPU and the plugin is reusing the buffer. If you move the object into the lambda it would mean you are not allowed to touch the buffer anymore.

Is that so? The way I see it, `Binner::notify()` in `include/picongpu/plugins/binning/Binner.hpp` allocates the buffer and the last use of it is in the call to `histWriter()`. Moving the buffer is the entire point of this optimization, so if we can't do that, then we can leave things as they are.

franzpoeschel replied:
> then we can leave things as they are.

.. we should at least use `storeChunkRaw()` instead of a fake shared pointer though

franzpoeschel replied:
If I read `include/pmacc/memory/buffers/HostBuffer.hpp` correctly, then `HostBuffer` does not implement move semantics? There is nothing that ever modifies `pointer` and `ownPointer`, so after a move you get a double free.
(If we don't want to implement those operations, it would probably be good to explicitly `delete` them, to avoid others falling into the same problem)

franzpoeschel replied:
[This commit](https://github.com/ComputationalRadiationPhysics/picongpu/commit/fe11baebaa840d7dd1a280e10b93f3c8389843ed) has the full suggestion for adding this optimization. Alternatively to the workaround there (putting the HostBuffer into a unique_ptr), one could also add move semantics to HostBuffer.
I've tested it, this one runs fine now.

franzpoeschel commented on include/picongpu/plugins/binning/WriteHist.hpp:
```diff
@@ -0,0 +1,171 @@
+                    filename << binningData.binnerOutputName << '.' << extension;
+                }
+
+                auto series = ::openPMD::Series(dir + '/' + filename.str(), ::openPMD::Access::CREATE);
```
Not necessary to do in this PR, but if at some point we want more than file-based encoding (one file per iteration), we should consider leaving the `Series` open between calls to this plugin.

ikbuibui committed:
Changed comment style, added EOLs
More small changes, added a openPMD include
Dont compile the example if electrons dont exist in the simulation

psychocoderHPC commented on include/picongpu/param/binningSetup.param:
```diff
@@ -0,0 +1,211 @@
+        }
+
+        template<typename T_Identifier>
+        struct NameOrTypeExists
```
Offline discussed we shoudl use `pmacc::particles::meta::FindByNameOrType_t` with the policy `ReturnType` or `ReturnValue`

ikbuibui replied:
Changed.

psychocoderHPC commented on include/picongpu/param/binningSetup.param:
```diff
@@ -0,0 +1,211 @@
+            static constexpr bool value = !std::is_same_v<boost::mp11::mp_true, mp_empty<FilteredSeq>>;
+        };
+
+        inline void binningExample(BinningCreator& binningCreator)
```
please add a comment that this is the check that the species exists. Best use

```
constexpr speciesExists = ....;
if constexpr(speciesExists) ...
```

ikbuibui replied:
Comment added

ikbuibui committed:
Use existing code to check for species
Remove NameOrTypeExists and use FindByNameOrType
